### PR TITLE
add linear functional

### PR DESCRIPTION
**Pull request by vieting**
Add `torch.nn.functional.linear` using the new module `FunctionalLinear`. This first needs the corresponding PR in returnn https://github.com/rwth-i6/returnn/pull/416

### CODE REVIEW

File: pytorch_to_returnn/torch/nn/modules/linear.py [hunk 1]
```diff
@@ -60,6 +60,27 @@ def import_params_torch_to_returnn(self, *, layer: LinearLayer, torch_module: Li
...
+    if bias is not None:
+      assert bias.shape == (out_features,)
+    return {
+      "class": "linear", "from": self._get_input_layer_name(input),
+      "n_out": out_features,
+      "with_bias": bias is not None,
+      "bias": self._get_input_layer_name(bias) if bias is not None else None,
+      "weights": self._get_input_layer_name(weight),
+      "use_transposed_weights": True,
+      "activation": None}
```
Comment (albertz) [inline pytorch_to_returnn/torch/nn/modules/linear.py:81]: Rather than defining `create_returnn_layer_dict`, you should define `forward` and use `functional.matmul` and just `+` for the bias.

File: pytorch_to_returnn/torch/nn/modules/linear.py [hunk 2]
```diff
@@ -60,6 +60,27 @@ def import_params_torch_to_returnn(self, *, layer: LinearLayer, torch_module: Li
       layer.params["b"].load(torch_module.bias.detach().numpy(), session=session)
 
 
+class FunctionalLinear(Module):
+  is_original_torch_module = False
+
+  def __init__(self):
+    super(FunctionalLinear, self).__init__()
+
```
Comment (albertz) [inline pytorch_to_returnn/torch/nn/modules/linear.py:68]:
```suggestion
```

Comment (vieting) [reply pytorch_to_returnn/torch/nn/modules/linear.py:68]: Is there a difference?

File: pytorch_to_returnn/torch/nn/modules/linear.py (hunk 1 above)
Comment (vieting) [reply pytorch_to_returnn/torch/nn/modules/linear.py:81]: True, but then actually we don't need this module at all, right? We can just do it in `functional.linear`.

File: pytorch_to_returnn/torch/nn/modules/linear.py (hunk 2 above)
Comment (albertz) [reply pytorch_to_returnn/torch/nn/modules/linear.py:68]: Functionality-wise, no, but it's cleaner.

File: pytorch_to_returnn/torch/nn/modules/linear.py (hunk 1 above)
Comment (albertz) [reply pytorch_to_returnn/torch/nn/modules/linear.py:81]: Yes. This is maybe a matter of taste. In one case, it will end up in a subnetwork, in the other not.

File: pytorch_to_returnn/torch/nn/functional.py [hunk 3]
```diff
@@ -197,6 +207,15 @@ def transpose(input: Tensor, dim0: int, dim1: int):
   return tensorflow_transpose(input, perm={dim0: dim1, dim1: dim0})
 
 
+def t(input: Tensor):
+  if len(input.shape) < 2:
+    return input
+  elif len(input.shape) == 2:
+    return transpose(input, 0, 1)
+  else:
+    raise ValueError("t() expects input to be <= 2-D tensor")
```
Comment (albertz) [inline pytorch_to_returnn/torch/nn/functional.py:216]: Is that the case? Or not implemented?

File: pytorch_to_returnn/torch/nn/modules/linear.py [hunk 4]
```diff
@@ -60,6 +60,42 @@ def import_params_torch_to_returnn(self, *, layer: LinearLayer, torch_module: Li
...
+      "class": "linear", "from": self._get_input_layer_name(input),
+      "n_out": out_features,
+      "with_bias": bias is not None,
+      "bias": self._get_input_layer_name(bias) if bias is not None else None,
+      "weights": self._get_input_layer_name(weight),
+      "use_transposed_weights": True,
+      "activation": None}
+
+
+class DotLayer(Module):
```
Comment (albertz) [inline pytorch_to_returnn/torch/nn/modules/linear.py:84]: I wonder whether we maybe should call it `Matmul` instead.

File: pytorch_to_returnn/torch/nn/modules/linear.py [hunk 5]
```diff
@@ -60,6 +60,42 @@ def import_params_torch_to_returnn(self, *, layer: LinearLayer, torch_module: Li
...
+  """
+  is_original_torch_module = False
+
+  def __init__(self):
+    super(DotLayer, self).__init__()
+
+  def create_returnn_layer_dict(self, *inputs: Tensor, **kwargs) -> Dict[str, Any]:
+    sources = [self._get_input_layer_name(source) for source in inputs]
+    assert len(sources) == 2
+    return {"class": "dot", "from": sources}
```
Comment (albertz) [inline pytorch_to_returnn/torch/nn/modules/linear.py:96]: This is probably incomplete, right? You definitely need to specify the axes.

File: pytorch_to_returnn/torch/nn/functional.py (hunk 3 above)
Comment (vieting) [reply pytorch_to_returnn/torch/nn/functional.py:216]: Yes, that's stated in the PyTorch docs https://pytorch.org/docs/stable/generated/torch.t.html#torch.t

File: pytorch_to_returnn/torch/nn/modules/linear.py (hunk 4 above)
Comment (vieting) [reply pytorch_to_returnn/torch/nn/modules/linear.py:84]: Yes, I guess that fits nicer into the PyTorch naming

File: pytorch_to_returnn/torch/nn/functional.py (hunk 3 above)
Comment (albertz) [reply pytorch_to_returnn/torch/nn/functional.py:216]: Ok. Maybe add a small comment. This is a bit unintuitive (because NumyPy and TensorFlow allow any N-D I think).

File: pytorch_to_returnn/torch/nn/modules/linear.py [hunk 6]
```diff
@@ -93,7 +93,11 @@ def __init__(self):
   def create_returnn_layer_dict(self, *inputs: Tensor, **kwargs) -> Dict[str, Any]:
     sources = [self._get_input_layer_name(source) for source in inputs]
     assert len(sources) == 2
-    return {"class": "dot", "from": sources}
+    red1 = inputs[0].returnn_naming_entry.get_returnn_axis_description(-1)
+    red2 = inputs[1].returnn_naming_entry.get_returnn_axis_description(-2)
+    var1 = inputs[0].returnn_naming_entry.get_returnn_axis_description(-2)
+    var2 = inputs[1].returnn_naming_entry.get_returnn_axis_description(-1)
```
Comment (albertz) [inline pytorch_to_returnn/torch/nn/modules/linear.py:99]: Better make this consistent as it is done in other layers. Via `self.get_axis...`

Comment (vieting) [reply pytorch_to_returnn/torch/nn/modules/linear.py:93]: I'm not sure what you mean here. I don't see modules which use a function `self.get_axis...`

File: pytorch_to_returnn/torch/nn/modules/linear.py [hunk 7]
```diff
@@ -60,6 +60,59 @@ def import_params_torch_to_returnn(self, *, layer: LinearLayer, torch_module: Li
...
+  def create_returnn_layer_dict(self, *inputs: Tensor, **kwargs) -> Dict[str, Any]:
+    sources = [self._get_input_layer_name(source) for source in inputs]
+    assert len(sources) == 2
+
+    assert len(inputs[0].shape) >= 2, "not implemented otherwise"
+    assert len(inputs[1].shape) >= 2, "not implemented otherwise"
+    red1 = inputs[0].returnn_naming_entry.get_returnn_axis_description(-1)
+    red2 = inputs[1].returnn_naming_entry.get_returnn_axis_description(-2)
+    var1 = [inputs[0].returnn_naming_entry.get_returnn_axis_description(-2)]
+    var2 = [inputs[1].returnn_naming_entry.get_returnn_axis_description(-1)]
```
Comment (albertz) [inline pytorch_to_returnn/torch/nn/modules/linear.py:96]: See other layers. Make this consistent. You can use `self._get_axis...` (or sth like that). Do not access `returnn_naming_entry`.

File: pytorch_to_returnn/torch/nn/modules/linear.py (hunk 1 above)
Comment (albertz) [reply pytorch_to_returnn/torch/nn/modules/linear.py:78]: This still should be cleaned up. I think you don't need it at all anymore, right?

File: pytorch_to_returnn/torch/nn/modules/linear.py (hunk 6 above)
Comment (albertz) [reply pytorch_to_returnn/torch/nn/modules/linear.py]: See how all other layers get the axis.

File: pytorch_to_returnn/torch/nn/modules/linear.py (hunk 7 above)
Comment (albertz) [reply pytorch_to_returnn/torch/nn/modules/linear.py:96]: `self._get_input_axis_to_returnn`

File: pytorch_to_returnn/torch/nn/modules/linear.py (hunk 6 above)
Comment (albertz) [reply pytorch_to_returnn/torch/nn/modules/linear.py]: `self._get_input_axis_to_returnn`

File: pytorch_to_returnn/torch/nn/modules/linear.py (hunk 1 above)
Comment (vieting) [reply pytorch_to_returnn/torch/nn/modules/linear.py]: That's right.

**Review by albertz — APPROVED** [review]
